agent/taskresource/cgroup/cgroup.go

```diff
@@ -0,0 +1,212 @@
+	defer c.lock.RUnlock()
+
+	return c.knownStatusUnsafe
+}
+
+// SetCreatedAt sets the timestamp for resource's creation time
+func (c *CgroupResource) SetCreatedAt(createdAt time.Time) {
+	if createdAt.IsZero() {
+		return
+	}
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.createdAt = createdAt
+}
+
+// GetCreatedAt sets the timestamp for resource's creation time
+func (c *CgroupResource) GetCreatedAt() time.Time {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.createdAt
+}
+
+// Create creates cgroup root for the task
+func (c *CgroupResource) Create(task *api.Task) error {
+	err := c.setupTaskCgroup(task)
+	if err != nil {
+		seelog.Criticalf("Unable to setup platform resources: %v", err)
+		return err
+	}
+	return nil
+}
+
+func (c *CgroupResource) setupTaskCgroup(task *api.Task) error {
+	cgroupRoot := c.CgroupRoot
+	seelog.Debugf("Setting up cgroup at: %s for task: %s", cgroupRoot, task.Arn)
+
+	if c.control.Exists(cgroupRoot) {
+		seelog.Debugf("Cgroup at %s already exists, skipping creation", cgroupRoot)
+		return nil
+	}
+
+	linuxResourceSpec, err := task.BuildLinuxResourceSpec()
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to build resource spec for task: %s", task.Arn)
+	}
+
+	cgroupSpec := cgroup.Spec{
+		Root:  cgroupRoot,
+		Specs: &linuxResourceSpec,
+	}
+
+	_, err = c.control.Create(&cgroupSpec)
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to create cgroup at %s for task: %s", cgroupRoot, task.Arn)
+	}
+
+	// enabling cgroup memory hierarchy by doing 'echo 1 > memory.use_hierarchy'
+	memoryHierarchyPath := filepath.Join(c.CgroupMountPath, memorySubsystem, cgroupRoot, memoryUseHierarchy)
+	err = c.ioutil.WriteFile(memoryHierarchyPath, enableMemoryHierarchy, rootReadOnlyPermissions)
+	if err != nil {
+		return errors.Wrap(err, "resource: setup cgroup: unable to set use hierarchy flag")
+	}
+
+	return nil
+}
+
+// Cleanup removes the cgroup root created for the task
+func (c *CgroupResource) Cleanup() error {
+	err := c.control.Remove(c.CgroupRoot)
+	// Explicitly handle cgroup deleted error
+	if err != nil {
+		if err == cgroups.ErrCgroupDeleted {
+			seelog.Warnf("Cgroup at %s has already been removed: %v", c.CgroupRoot, err)
+			return nil
+		}
+		return errors.Wrapf(err, "resource: cleanup cgroup: unable to remove cgroup at %s", c.CgroupRoot)
+	}
+	return nil
+}
+
+// cgroupResourceJSON duplicates CgroupResource fields, only for marshalling and unmarshalling purposes
+type cgroupResourceJSON struct {
+	CgroupRoot      string        `json:"CgroupRoot"`
+	CgroupMountPath string        `json:"CgroupMountPath"`
+	CreatedAt       time.Time     `json:",omitempty"`
+	DesiredStatus   *CgroupStatus `json:"DesiredStatus"`
+	KnownStatus     *CgroupStatus `json:"KnownStatus"`
+}
+
+// MarshalJSON marshals CgroupResource object using duplicate struct CgroupResourceJSON
+func (c *CgroupResource) MarshalJSON() ([]byte, error) {
+	return json.Marshal(cgroupResourceJSON{
```

Comment:
You need to do a `nil` check here. Example: https://github.com/aws/amazon-ecs-agent/blob/f4d6e452cf23a15cdc94cc3ee9f1ee5fb27679e5/agent/taskresource/volume.go#L114

Reply:
Sure. I don't understand why go lets you call a method on a nil object.

Reply:
https://groups.google.com/forum/#!topic/golang-nuts/wcrZ3P1zeAk

Reply:
Looks like go is really confused between being an non-object oriented and object oriented language. 😄